public/app/plugins/panel/barchart/utils.ts

```diff
@@ -493,7 +493,18 @@ export function prepareBarChartDisplayValues(
   let legendFields: Field[] = fields;
   if (options.stacking === StackingMode.Percent) {
     legendFields = fields.map((field) => {
-      const alignedFrameField = frame.fields.find((f) => f.state?.displayName === field.state?.displayName)!;
+      let alignedFrameField: Field;
+
+      // if displayNameFromDS is set, use that to align the legend fields, otherwise use the displayName
+      if (field.config.displayNameFromDS) {
+        alignedFrameField = frame.fields.find(
+          (frameField) => frameField.config.displayNameFromDS === field.config.displayNameFromDS
+        )!;
+      } else {
+        alignedFrameField = frame.fields.find(
+          (frameField) => frameField.state?.displayName === field.state?.displayName
+        )!;
+      }
```

Comment:
hm.. strange, field.state.displayName should have the displayNameFromDS value, 
https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/field/fieldState.ts#L85 

is it something in the stacking logic that changes the field somehow, is the state cleared? what if we called getFieldDisplayName here instead?

Reply:
I think the state is being cleared on line 499 mainly to clear out the field config unit to prevent legend values from being percents (see [this related PR](https://github.com/grafana/grafana/pull/61449))

Tested out using the `getFieldDisplayName` function and that resolves the issue / seems like a cleaner approach than this if branching :)

Reply:
Update on using `getFieldDisplayName` - it seems to break field name processing in a strange way. I need to investigate this further when I'm back next week after holidays.

The breakage occurs when renaming the second query value (see the screencast)


https://github.com/grafana/grafana/assets/22381771/750a491d-4276-4972-990b-d971982cb7b7



[debug-Stacking 100% Correct Legend Units (no %)-2023-11-23 10_31_39.json.txt](https://github.com/grafana/grafana/files/13452850/debug-Stacking.100.Correct.Legend.Units.no.-2023-11-23.10_31_39.json.txt)

Reply:
resolved issue - fix was found by [uncovering root changes](https://github.com/grafana/grafana/pull/61449) that are causing the recent instability around this feature and rethinking a different, more stable approach.